docs/rfc/rfc-016-node-architecture.md

```diff
@@ -0,0 +1,83 @@
+
+The schematic supplied above shows the relations between the different services, the node, the block executor, and the storage layer. Represented as colourful dots are the components responsible for different roles of interprocess communication (IPC). These components permeate throughout the code base, seeping into most services. What can provide powerful functionality on one hand can also become a twisted vine, creating messy corner cases and convoluting the protocols themselves. A lot of the thinking around
+how we want our IPC systens to function has been summarised in this [RFC](./rfc-002-ipc-ecosystem.md). In this sectionm, I'd like to focus the reader on the relation between the IPC and the node structure. An issue that has frequently risen is that the RPC has control of the components where it strikes me as being more logical for the component to dictate the information that is emitted/available and the knobs it wishes to expose. The RPC is also inextricably tied to the node instance and has situations where it is passed pointers directly to the storage engine and other components.
+
+I am currently convinced of the approach that the p2p layer takes and would like to see other IPC components follow suit. This would mean that the RPC and events system would be constructed in the node yet would pass the adequate methods to register endpoints and topics to the sub components. For example,
+
+```go
+// Methods from the RPC and event bus that would be passed into the constructor of components like "consensus"
+// NOTE: This is a hypothetical construction to convey the idea. An actual implementation may differ.
+func RegisterRoute(path string, handler func(http.ResponseWriter, *http.Request))
+
+func RegisterTopic(name string) EventPublisher
+
+type EventPublisher func (context.Context, types.EventData, []abci.Event)
+```
+
+This would give the components control to the information they want to expose and keep all relevant logic within that package. It accomodates more to a dynamic system where services can switch on and off. Each component would also receive access to the logger and metrics system for introspection and debuggability.
+
+#### IPC Rubric
+
+I'd like to aim to reach a state where we as a team have either an implicit or explicit rubric which can determine, in the event of some new need to communicate information, what tool it should use for doing this. In the case of inter node communication, this is obviously the p2p stack (with perhaps the exception of the light client). Metrics and logging also have clear usage patterns. RPC and the events system are less clear. The RPC is used for debugging data and fine tuned operator control as it is for general public querying and transaction submission. The RPC is also known to have been plumbed back into the application for historical queries. The events system, similarly, is used for consuming transaction events as it is for the testing of consensus state transitions.
```

Comment:
Why is the light client necessarily an exception here? Is that a fundamental requirement or an artifact of the way it's implemented now? 

I ask mainly because I know we don't have full design control over light client APIs, due to their use in IBC.

Reply:
An artifact of the way it's implemented. I would consider the light client a type of node and should be part of the p2p stack. One of the main reasons for building it on top of the RPC was that it would be easier to support light client implementations in other languages (hopefully libp2p remedies that somewhat)